controllers/helper_test.go

```diff
@@ -538,6 +540,52 @@ func testBuildManagedZone(name, ns, domainName string) *kuadrantdnsv1alpha1.Mana
 	}
 }
 
+// testEndpointsTraversable consumes an array of endpoints and returns a boolean
+// indicating presence of that path from host to all IPs
+func testEndpointsTraversable(endpoints []*endpoint.Endpoint, host string, ips []string) bool {
+	for _, ip := range ips {
+		allTargetsFound := false
+		for _, ep := range endpoints {
+			// the target on the endpoint exists as an endpoint
+			if ep.DNSName == host {
+				// target is our IP
+				if slices.Contains(ep.Targets, ip) {
+					return true
+				} else {
+					// target is not out IP. Use target as a host and check for existence of Endpoints with such a DNSName
+					for _, target := range ep.Targets {
+						// if at least one returns as true allTargetsFound will be locked in true
+						allTargetsFound = allTargetsFound || testEndpointsTraversable(endpoints, target, []string{ip})
+					}
+				}
+			}
+		}
+		// at least one endpoint that matches host targeted IP
+		return allTargetsFound
+	}
+	// there are no IPs to reach / default response
+	return false
+}
+
+// testDNSRecordsTraversable consumes a list of DNSRecords and tests if at least one of them could
+// be traversable from any of key/value pairs: hostName (the key) to all the IPs (the []string value)
+// returns false if a key/value pair is not traversable on any of the DNSRecords in the list
+// returns true if all key/value pairs are traversable on at least one records in the list
+func testDNSRecordsTraversable(list *kuadrantdnsv1alpha1.DNSRecordList, hostnames map[string][]string) bool {
+	for hostname, IP := range hostnames {
+		for _, record := range list.Items {
+			// found traversable record
+			if testEndpointsTraversable(record.Spec.Endpoints, hostname, IP) {
+				return true
+			}
+		}
+		// no records are traversable for this key/value pair
+		return false
+	}
+	// no hostnames to check
+	return false
+}
```

Comment:
some of the tests deal with the list of records. It is either this or adjusting some of the tests. This check is not expensive so i went with this approach to allow for more flexibility in the tests. (and i just spotted that if the first record is traversable this func returns `true` without checking the second).

Reply:
I would probably have changed the tests tbh. If this started failing it will probably be hard enough to debug the issue without having arrays of records you need to look at.

Originally we had been doing assertions on the arrays of records, and there are obviously still some tests that are, but we should really change those since it was proving difficult to debug.

Reply:
Rewrote tests. Now only one way of checking traversability